src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -1983,64 +1997,61 @@ else if (compareAndSetCtl(c, c) && casRunState(e, e | STOP))
+                                        a[(b + 2) & m] == null) {
+                                        break;            // probably empty
+                                    }
+                                    if (pb == (pb = b)) { // track progress
+                                        rescan = true;    // stalled; reorder scan
+                                        break scan;
                                     }
-                                    else if (taken)
-                                        continue rescan;  // depleted; restart
-                                    else
-                                        break;            // empty
                                 }
-                                if (pb == (pb = b))       // base unchanged
-                                    Thread.onSpinWait();
                             }
-                            else if (!U.compareAndSetReference(a, kp, t, null))
-                                b = q.base;               // contended
                             else {
-                                q.base = nb;
+                                boolean propagate;
+                                int nb = q.base = b + 1;
                                 w.nsteals = ++nsteals;
-                                w.source = j;             // volatile write
-                                if (taken != (taken = true) && a[nk] != null)
-                                    signalWork();         // propagate signal
-                                w.topLevelExec(t, cfg);
-                                if ((b = q.base) != nb && src != (src = j))
-                                    continue rescan;      // reduce interference
+                                w.source = j;             // volatile
+                                rescan = true;
+                                if (propagate =
+                                    ((src != (src = j) ||
+                                      t.getClass().getSuperclass() ==
+                                      interruptibleTaskClass) &&
+                                     a[nb & m] != null))
+                                    signalWork();
+                                w.topLevelExec(t, fifo);
+                                if ((b = q.base) != nb && !propagate)
+                                    break scan;          // reduce interference
                             }
                         }
                     }
                 }
-                if (!scanned)
-                    scanned = true;                   // rescan before deactivate
-                else if (((phase = deactivate(w, r, phase)) & IDLE) == 0)
-                    scanned = false;
-                else
-                    return;
+                if (!rescan) {
+                    if (((phase = deactivate(w, phase)) & IDLE) != 0)
+                        break;
+                    src = -1;
```

Comment:
@DougLea So if I understand this correctly, the setting of `src` to -1 is to differ from `j` on the next loop step (since `j` is bound by bitmask to be positive)?

Reply:
yes. To be little clearer added:
                    src = -1;                            // re-enable propagation